### PR TITLE
[CodeCompletion] Stop using temporary Lexer in the second pass

### PR DESCRIPTION
**Pull request by rintaro**
* Since we only call one parsing function (i.e. `parseBraceItemList`, `parseDecl`, or `parseExprOrStmt`), the parser stops at the end of the node anyways. It's not necessary to limit the Lexer to set an ArtificialEOF. As a ground work, modify the Parser to *not* parse the body if the completion happens in the signature
* Don't restore the parser position after the second pass because it's just not necessary. The parser here lives only for the single second pass.
* Use `getResultInterfaceType()` to get the result type in `TypeCheckFunctionBodyUntilRequest`. So we can type check the body without type checking the signature of the func decl. (This change is actually not needed but I think this is an improvement)

### CODE REVIEW

File: lib/Sema/TypeCheckStmt.cpp [hunk 1]
```diff
@@ -2121,15 +2121,11 @@ TypeCheckFunctionBodyUntilRequest::evaluate(Evaluator &evaluator,
...
-      if (resultTypeLoc.isNull() || resultTypeLoc.getType()->isVoid()) {
-        // The function returns void.  We don't need an explicit return, no matter
-        // what the type of the expression is.  Take the inserted return back out.
-        body->setFirstElement(expr);
-      }
+    } else if (func->hasSingleExpressionBody() &&
+               func->getResultInterfaceType()->isVoid()) {
+      // The function returns void.  We don't need an explicit return, no matter
+      // what the type of the expression is.  Take the inserted return back out.
+      body->setFirstElement(func->getSingleExpressionBody());
```
Comment (rintaro) [inline lib/Sema/TypeCheckStmt.cpp]: @DougGregor Do you have any concern about this change?

File: test/Parse/recovery.swift [hunk 2]
```diff
@@ -695,9 +695,6 @@ struct InitializerWithNameAndParam {
 struct InitializerWithLabels {
   init c d: Int {}
   // expected-error @-1 {{expected '(' for initializer parameters}}
-  // expected-error @-2 {{expected declaration}}
-  // expected-error @-3 {{consecutive declarations on a line must be separated by ';'}}
-  // expected-note @-5 {{in declaration of 'InitializerWithLabels'}}
```
Comment (rintaro) [inline test/Parse/recovery.swift]: This is because `parseDeclInit` now returns an error state. So the parser property skips to the next decl or '}'

**Comment by rintaro** [comment]
@swift-ci Please smoke test

File: lib/Sema/TypeCheckStmt.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/Sema/TypeCheckStmt.cpp]: No, this seems like a clear improvement.

**Review by DougGregor — APPROVED** [review]
Nice cleanup!

**Review by benlangmuir — APPROVED** [review]

**Pull request merged**